ouroboros-network/src/Ouroboros/Network/Connections/Socket/Types.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Ouroboros.Network.Connections.Socket.Types
+  ( SockType (..)
+  , SockAddr (..)
+  , ConnectionId (..)
+  , makeConnectionId
+  , connectionIdPair
+  , Some (..)
+  , someSockType
+  , withSockType
+  , matchSockType
+  , matchSockAddr
+  , forgetSockType
+  ) where
+
+import Data.Kind (Type)
+import qualified Network.Socket as Socket
+
+data SockType where
+  IPv4 :: SockType
+  IPv6 :: SockType
+  Unix :: SockType
+
+-- | Like `Network.Socket.SockAddr` but with a type parameter indicating which
+-- kind of address it is: IPv4, IPv6, or Unix domain.
+data SockAddr (sockType :: SockType) where
+  SockAddrIPv4 :: !Socket.PortNumber -> !Socket.HostAddress -> SockAddr IPv4
+  SockAddrIPv6 :: !Socket.PortNumber
+               -> !Socket.FlowInfo
+               -> !Socket.HostAddress6
+               -> !Socket.ScopeID
+               -> SockAddr IPv6
+  -- No strictness here because the Network.Socket analogue is not strict.
+  SockAddrUnix :: String -> SockAddr Unix
```

Comment:
hmm, so why laziness here is a good choice?

Reply:
Just because that's how the analogue type in `network` does it. It'd probably be fine to make it strict. But also, not really a big deal IMO since at worst we keep around 4 unevaluated thunks that are probably not very big. i.e. the computation of the host or port is probably not a very complicated closure.

Reply:
Open to making it strict. I am indifferent.